### ISSUE TITLE
Better defaults to match Hugging Face

### ISSUE THREAD

**Issue by titu1994** [vllm-project/vllm#2733]
While I appreciate the fact that CUDA kernels can cause issues with bitwize exactness, there's an issue with vLLM defaults that causes significantly worse results on benchmarks than using the corresponding slow Huggingface code. 

This becomes a much bigger deal with using the openai api_endpoint.py script - the lack of easy way to know the defaults of vllm + how to pass arguments that arent normally accepted by the completions API caused me to waste practically an entire day. Hopefully the solution below helps someone not waste that much time.

So in an effort to enable at least greedy scores to be closely aligned, here is some example code that helps achieve nearly same greedy scores on HumanEval compared to HF but about 20x faster.

# Serve model

(Note: as of v0.3.0 there a bug with logprobs calculation and using echo=True). The fix is easy but will need manual patching of the `serving_completion.py` script. We also use a large max-num-seqs cause we want to process entire dataset in one pass if memory allows. I'll use a mistral 7b instruct 0.1 for this example.

```python
python -m vllm.entrypoints.openai.api_server \
      --model "${MODEL_PATH}" \
      --trust-remote-code \
      --seed=${SEED} \
      --host="0.0.0.0" \
      --port=${SERVER_PORT} \
      --served-model-name "${MODEL_NAME}" \
      --tensor-parallel-size=${NUM_GPUS} \
      --max-num-seqs=1024
```

# Generate solutions for Humaneval

The main parameter difference between HF and vLLM is `top_k` - HF's default for greedy is 1 and for `do_sample=True` is 50 (odd choice but ok). vLLM on the other hand has an obscure `-1` as the default (I presume it means it is disabled and all logits are used?). Well here starts the problem. If you use vLLM endpoint, sure you can explicitly set `top_k` to 1 or 50, but openai endpoint is generally used with the openai library - and that doesn't allow passing top_k.

You gotta hunt down the docs for an obscure flag `extra_body` which accepts a dict as additional arguments to pass into the request. Oh well.

Another weird flag is `spaces_between_special_tokens`- it has a weak but negative effect on scores when set to True by default. Maybe useful in other non-code benchmarks perhaps.

The crucial part of the script below is 

```python
    # Setup top k properly (to match HF)
    if generation_config['top_k'] >= 1:
        extra_body = {'top_k': generation_config['top_k'], 'spaces_between_special_tokens': False}
    else:
        extra_body = {'spaces_between_special_tokens': False}
```

and

```python
 response = client.completions.create(
                model=model_name,
                prompt=prompt_subbatch,
                max_tokens=generation_config['tokens_to_generate'],
                temperature=generation_config['temperature'],
                top_p=generation_config['top_p'],
                n=args.N,
                stream=False,
                stop=generation_config['end_strings'],
                echo=True,
                frequency_penalty=0.0,
                presence_penalty=0.0,
                extra_body=extra_body,
            )
```

## Script to generate solutions 

```python
"""
Mistral 7b
python temp.py \
    --output_path "results/mistral_7b_instruct_humaneval_vllm" \
    --N 1 --greedy_decode --batchsize 165 \
    --temperature 1.0 --top_p 1.0 --max_len 1280

"""

import argparse
import pprint
import os
from tqdm import tqdm

import torch
import openai
from evalplus.data import get_human_eval_plus, write_jsonl


def generate_prompt(input):
    INSTRUCTION = f"""[INST] Create a Python script for this problem:
{input} [/INST]

### Response:"""
    return INSTRUCTION


if torch.cuda.is_available():
    device = "cuda"
else:
    device = "cpu"


def main():
    parser = argparse.ArgumentParser()

    parser.add_argument('--host', type=str, default='127.0.0.1', help='')
    parser.add_argument('--port', type=int, default=5000, help='')
    parser.add_argument('--output_path', type=str, help="")
    parser.add_argument('--start_index', type=int, default=0, help="")
    parser.add_argument('--end_index', type=int, default=164, help="")
    parser.add_argument('--temperature', type=float, default=1.0, help="")
    parser.add_argument('--top_p', type=float, default=1.0, help="")
    parser.add_argument('--N', type=int, default=1, help="")
    parser.add_argument('--max_len', type=int, default=1280, help="")
    parser.add_argument('--greedy_decode', action='store_true', help='')
    parser.add_argument('--batchsize', type=int, default=50, help='')
    parser.add_argument('--overwrite', action='store_true', help='')

    # Model specification
    args = parser.parse_args()

    argsdict = vars(args)
    print(pprint.pformat(argsdict))

    problems = get_human_eval_plus()

    task_ids = sorted(problems.keys())[args.start_index: args.end_index]
    prompts = [problems[task_id]['prompt'] for task_id in task_ids]
    num_samples = len(prompts)
    print("Number of samples: {}".format(num_samples))

    generation_config = {
        "tokens_to_generate": args.max_len,
        "temperature": args.temperature,
        "top_k": 1 if args.greedy_decode else 50,  # match HF
        "top_p": args.top_p,
        "greedy": True if args.greedy_decode else False,
        'end_strings': ['</s>', 'Explain:', 'Explanation:', 'Example:'],  # Early exit the explanations, not needed
    }

    if generation_config['greedy']:
        print("Greedy generation set - setting temperature, top_p and top_k to 1.0")
        generation_config['temperature'] = 1.0
        generation_config['top_p'] = 1.0
        generation_config['top_k'] = 1

    # Setup OpenAI
    client = openai.OpenAI(api_key='EMPTY', base_url=f"http://{args.host}:{args.port}/v1")
    model_list = client.models.list()
    model_name = model_list.data[0].id
    print("Model name:", model_name)

    # Setup top k properly (to match HF)
    if generation_config['top_k'] >= 1:
        extra_body = {'top_k': generation_config['top_k'], 'spaces_between_special_tokens': False}
    else:
        extra_body = {'spaces_between_special_tokens': False}

    basedir = os.path.expanduser(args.output_path)
    if not os.path.exists(basedir):
        os.makedirs(basedir, exist_ok=True)

    # Greedy batch generation
    if generation_config['greedy']:
        print("Generating completions for greedy decoding...")
        prog = tqdm(range(num_samples), ncols=0, total=num_samples)
        for i in range(0, num_samples, args.batchsize):
            output_files = []
            for j in range(args.batchsize):
                if i + j >= args.end_index:
                    break

                output_file = args.output_path + '/{}.jsonl'.format(args.start_index + i + j)

                if os.path.exists(output_file) and not args.overwrite:
                    print(f'Skip {output_file} as it already exists')
                    continue

                output_files.append(output_file)

            if len(output_files) == 0:
                continue

            prompt_subbatch = []
            ids_subbatch = []
            for j in range(args.batchsize):
                if i + j < args.end_index:
                    prompt = prompts[i+j].replace('    ', '\t')
                    prompt_subbatch.append(generate_prompt(prompt))

                    idx = task_ids[i + j]
                    ids_subbatch.append(idx)

            response = client.completions.create(
                model=model_name,
                prompt=prompt_subbatch,
                max_tokens=generation_config['tokens_to_generate'],
                temperature=generation_config['temperature'],
                top_p=generation_config['top_p'],
                n=args.N,
                stream=False,
                stop=generation_config['end_strings'],
                echo=True,
                frequency_penalty=0.0,
                presence_penalty=0.0,
                extra_body=extra_body,
            )

            # Extract completions
            gen_seqs = []
            for choice in response.choices:
                gen_seqs.append(choice.text)

            if gen_seqs is not None:
                for k, gen_seq in enumerate(gen_seqs):
                    idx = ids_subbatch[k]
                    output_file = output_files[k]

                    completion_seqs = []
                    completion_seq = gen_seq[len(prompt_subbatch[k]):]
                    completion_seq = completion_seq.replace('\t', '    ')
                    all_code = gen_seq.replace('\t', '    ')

                    completion_seqs.append(
                        {'task_id': idx,
                         'completion': completion_seq,
                         'all_code': all_code,
                         }
                    )

                    print("Saving results to {}".format(output_file))
                    write_jsonl(output_file, completion_seqs)

                print()
                prog.update(len(gen_seqs))

        print("Finished!")
        return


if __name__ == '__main__':
    main()
```

----

Thats about it, this two little flags allow pretty much 99% similar greedy eval for huggingface with the openai endpoint with the openai library.

Id suggest if possible to add a short note about the top_k flag and how to pass it into openai as a note somewhere in the docs, so that others in he future don't have to hunt down these arguments like i did.

Fyi @simon-mo @zhuohan123

**Comment by zhuohan123** [comment]
> For HF compatibility, I think it's worth it to set these as the new default. Thoughts? @WoosukKwon @zhuohan123 @Yard1

I think before we try to set the default as the API provided by OpenAI. Which one do you think is better? OpenAI or HF?

**Comment by simon-mo** [comment]
OpenAI doesn't allow `top_k`

**Comment by Yard1** [comment]
I think we should standardize on OpenAI too. This is something that should be documented, though!

**Comment by njhill** [comment]
Presumably superset of OpenAI though? Since there are a lot of restrictions/missing options

**Comment by Yard1** [comment]
Yes, we should provide a superset of OpenAI parameters/features, but we should use OpenAI's default values.

**Comment by hmellor** [comment]
@Yard1 do you know if anyone has documented this yet?

**Comment by hmellor** [comment]
Default temperature has been changed from 0.7 to 1.0, which matches HF and OpenAI https://github.com/vllm-project/vllm/pull/11219

**Comment by titu1994** [comment]
@hmellor Just a heads up, temperature has little to no impact on matching hugging faces scores in this thread. You can verify this by changing temperature to 1.0 in the above provided code, but removing the top k extra body. Your benchmark score of models will Not match (be worse than) greedy HF scores.

**Comment by hmellor** [comment]
Ah I see, sorry for the misunderstanding.

I've looked into this a little deeper and it seems as though the slow tokenizers have `spaces_between_special_tokens: bool = True` in [_decode](https://github.com/huggingface/transformers/blob/2c47618c1a282f925446506d53108dc6e82d9ef0/src/transformers/tokenization_utils.py#L1081), but the fast tokenizers do not have this argument. In some [old documentation](https://huggingface.co/transformers/v3.4.0/main_classes/tokenizer.html#transformers.PreTrainedTokenizer.decode) for v3.4.0 it states that the default for the fast tokenizers is `False`.

This likely explains the discrepancy because vLLM's default `tokenizer_mode = 'auto'` defaults to fast tokenizers. Therefore, for this argument the default value for `spaces_between_special_tokens` unfortunately needs to depend on `tokenizer_mode` in order to match HF 1:1.

As for `top_k` it is documented in https://docs.vllm.ai/en/latest/dev/sampling_params.html, is that sufficient to resolve your original issue?

**Comment by titu1994** [comment]
Thanks for reopening. A note right after the first example of using OpenAI API on that page would be good. The thing is, keeping it on the bottom is already after you visit the chat and completion section, you'll probably not browse those sections since people see it's compatible with the OpenAI API by then.

**Comment by titu1994** [comment]
About the extra space between tokenizer, it's effect is marginal. It doesn't affect scores a lot.
